cel/library.go

```diff
@@ -439,135 +438,1241 @@ func (l library) validateEmail(addr string) bool {
+			dotted := newIpv4(i.dottedRaw)
+			if dotted.address() {
+				i.dottedAddr = dotted
+				return true
+			}
+			return false
+		}
+		if i.h16() {
+			continue
+		}
+		if i.take(':') { //nolint:nestif
+			if i.take(':') {
+				if i.doubleColonSeen {
+					return false
+				}
+				i.doubleColonSeen = true
+				i.doubleColonAt = len(i.pieces)
+				if i.take(':') {
+					return false
+				}
+			}
+			continue
+		}
+		if i.str[i.index] == '%' && !i.zoneID() {
+			return false
+		}
+		break
+	}
+	return i.doubleColonSeen || len(i.pieces) == 8
+}
+
+// There is no definition for the character set allowed in the zone
+// identifier. RFC 4007 permits basically any non-null string.
+//
+// RFC 6874 defines it as ZoneID = 1*( unreserved / pct-encoded ).
+func (i *ipv6) zoneID() bool {
+	start := i.index
+	if i.take('%') {
+		if i.strLen-i.index > 0 {
+			// permit any non-null string
+			i.index = i.strLen
+			i.zoneIDFound = true
+			return true
+		}
+	}
+	i.index = start
+	i.zoneIDFound = false
+	return false
+}
+
+// Parses str from the current index to determine if it contains a valid
+// dotted sequence defined as:
+//
+// 1*3DIGIT "." 1*3DIGIT "." 1*3DIGIT "." 1*3DIGIT
+// Stores match in `dottedRaw`.
+func (i *ipv6) dotted() bool {
+	start := i.index
+	i.dottedRaw = ""
+	for {
+		if i.index < i.strLen && (i.digit() || i.take('.')) {
+			continue
+		}
+		break
+	}
+	if i.index-start >= 7 {
+		i.dottedRaw = i.str[start:i.index]
+		return true
+	}
+	i.index = start
+	return false
+}
+
+// h16 = 1*4HEXDIG
+// Stores 16-bit value in `pieces`.
+func (i *ipv6) h16() bool {
+	start := i.index
+	for {
+		if i.index >= i.strLen || !i.hexdig() {
+			break
+		}
+	}
+	str := i.str[start:i.index]
+	if len(str) == 0 {
+		// too short
+		return false
+	}
+	if len(str) > 4 {
+		// too long
+		return false
+	}
+
+	value, err := strconv.ParseUint(str, 16, 16)
+	if err != nil {
+		return false
+	}
+	i.pieces = append(i.pieces, uint16(value))
+	return true
+}
+
+// HEXDIG =  DIGIT / "A" / "B" / "C" / "D" / "E" / "F".
+func (i *ipv6) hexdig() bool {
+	c := i.str[i.index]
+	if ('0' <= c && c <= '9') ||
+		('a' <= c && c <= 'f') ||
+		('A' <= c && c <= 'F') {
+		i.index++
+		return true
+	}
+	return false
+}
+
+// DIGIT = %x30-39  ; 0-9.
+func (i *ipv6) digit() bool {
+	c := i.str[i.index]
+	if '0' <= c && c <= '9' {
+		i.index++
+		return true
+	}
+	return false
+}
+
+// If char is at the current index, return true and increment the index.
+// If char is not at the current index or the end of str has been reached,
+// return false.
+func (i *ipv6) take(char byte) bool {
+	if i.index >= i.strLen {
+		return false
+	}
+	if i.str[i.index] == char {
+		i.index++
+		return true
+	}
+	return false
+}
+
+// newIpv6 creates a new ipv6 based on str.
+func newIpv6(str string) *ipv6 {
+	return &ipv6{
+		str:           str,
+		strLen:        len(str),
+		pieces:        make([]uint16, 0),
+		doubleColonAt: -1,
+		dottedAddr:    nil,
```

Comment:
Can't we remove these constructor functions?

- Since go strings are immutable, I suspect that storing `strLen` doesn't give us any benefit over `len(i.str)`.
- I'm pretty we can just let `pieces` be a nil slice.
- We have a separate bool to track whether we've seen a double colon, so `doubleColonAt` doesn't need to be initialized to `-1`. 
- `dottedAddr: nil` is redundant.

This only leaves `str`. Which means it's feasible to use the struct without a constructor function:

```diff
-		dotted := newIpv4(i.dottedRaw)
+		dotted := &ipv4{str: i.dottedRaw}
```

@jchadwick-buf, @rodaine WDYT?

Reply:
`strLen` is indeed unnecessary, and `pieces` should be left as a nil slice. The only one I'd still set is `doubleColonoAtr` to avoid having invalid state (even if there is an external boolean tracking the colon's presence).

Reply:
I fixed up the initialization and removed the constructor function for `NewUri`, but I left the other two constructor functions. I can remove them if you feel strongly about it, but I just think the code is much cleaner and reads much nicer with them for the IPv4 and IPv6 addresses. Plus, I'm not sure it really buys us anything removing them other than a few LOC?

Reply:
:+1: I'll die on another hill.